notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb

```diff
@@ -0,0 +1,838 @@
+    "\n",
+    "s3_train_data = 's3://{}/{}'.format(BUCKET, train_prefix)\n",
+    "output_path = 's3://{}/{}'.format(BUCKET, output_prefix)\n",
+    "\n",
+    "print('Training set location', s3_train_data)\n",
+    "print('Trained model will be saved at', output_path)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def split_convert_upload(sparray, bucket, prefix, fname_template='data_part{}.pbr', n_parts=2):\n",
+    "    import io\n",
+    "    import sagemaker.amazon.common as smac\n",
+    "    \n",
+    "    chunk_size = sparray.shape[0]// n_parts\n",
+    "    for i in range(n_parts):\n",
+    "\n",
+    "        # Calculate start and end indices\n",
+    "        start = i*chunk_size\n",
+    "        end = (i+1)*chunk_size\n",
+    "        if i+1 == n_parts:\n",
+    "            end = sparray.shape[0]\n",
+    "        \n",
+    "        # Convert to record protobuf\n",
+    "        buf = io.BytesIO()\n",
+    "        smac.write_spmatrix_to_sparse_tensor(array=sparray[start:end], file=buf, labels=None)\n",
+    "        buf.seek(0)\n",
+    "        \n",
+    "        # Upload to s3 location specified by bucket and prefix\n",
+    "        fname = os.path.join(prefix, fname_template.format(i)).replace(\"\\\\\",\"/\")\n",
+    "        boto3.resource('s3').Bucket(bucket).Object(fname).upload_fileobj(buf)\n",
+    "        print('Uploaded data to s3://{}'.format(os.path.join(bucket, fname).replace(\"\\\\\",\"/\")))\n",
+    "split_convert_upload(vectors, bucket=BUCKET, prefix=train_prefix, fname_template='train_part{}.pbr', n_parts=1)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Load the latest NTM container"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sagemaker.amazon.amazon_estimator import get_image_uri\n",
+    "container = get_image_uri(boto3.Session().region_name, 'ntm', 'latest')\n",
+    "print(container)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Model Training\n",
+    "\n",
+    "To train the model, you can use 1 or more instances (specified by train_instance_count) and choose a strategy to either Fully Replicate the data on each instance or use ShardedByS3Key which only puts certain data shards on each instance, thus speeding up the training at the cost of each instance only seeing a fraction of the data.\n",
+    "\n",
+    "Another way to avoid overfitting is to introduce EarlyStopping which is done here by num_patience_epochs which ensure that the training is stopped if the change in the loss is less than the specified tolerance for a number of epochs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%time\n",
```

Comment:
If this magic command is intended to measure the time required to execute this cell, should use "%%time" instead.